OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+  ///                      the last col only partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws OpenSim::ZeroElements If the number of elements produced by the
+  ///                               input iterator is zero.
+  /// \throws OpenSim::InvalidEntry If the function is called on an empty
+  ///                               DataTable_ without providing the argument
+  ///                               nrow.
+  /// \throws OpenSim::InvalidEntry If allow_missing is false and the input
+  ///                               iterator does not fill up the last col
+  ///                               completely.
+  template<typename InputIt>
+  void addCols(InputIt first, 
+               InputIt last, 
+               const size_t nrow = std::numeric_limits<size_t>::max(),
+               const bool allow_missing = false) {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if((data.nrow() == 0 || data.ncol() == 0) && 
+       (nrow == std::numeric_limits<size_t>::max() || nrow == 0))
+      throw InvalidEntry{"DataTable is empty. 'nrow' argument must be" 
+                         " provided and it cannot be zero."};
+
+    data.resizeKeep(data.nrow() == 0 ? static_cast<int>(nrow) : data.nrow(), 
+                    data.ncol() + 1);
+    int row{0};
+    int col{data.ncol() - 1};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first; ++row;
+      if(row == static_cast<int>(data.nrow()) && first != last) {
+        row = 0;
+        ++col;
+        data.resizeKeep(data.nrow(), data.ncol() + 1);
+      }
+    }
+    if(!allow_missing && row != data.nrow())
+      throw InvalidEntry{"Input iterator did not produce enough elements to "
+                         "fill the last col."};
+  }
+
+  /// Bind another DataTable_ to this DataTable_ by row. The new elements will 
+  /// appear as the last rows of this DataTable_. Only data will be appended.
+  /// Metadata is not added. Columns retain their labels. To create a new 
+  /// DataTable_  that is a bind of two existing DataTable_(s), see 
+  /// rbindDataTables().
+  ///
+  /// \throws OpenSim::InvalidEntry If input DataTable_ has incorrect number of
+  ///                               columns for bind to work.
+  /// \throws Opensim::InvalidEntry If trying to bind a DataTable_ to itself.
+  void rbindDataTable(const DataTable_& table) {
+    if(data.ncol() != table.data.ncol()) 
+      throw InvalidEntry{"Input DataTable has incorrect number of columns."};
+    if(&data == &table.data)
+      throw InvalidEntry{"Cannot rbind a DataTable to itself."};
+
+    int old_nrow{data.nrow()};
+    data.resizeKeep(data.nrow() + table.data.nrow(), data.ncol());
+    data.updBlock(old_nrow         ,           0, 
+                  table.data.nrow(), data.ncol()).updAsMatrix() = table.data;
```

Comment:
BTW, in case you like it better there is a compact operator form of matrix block extraction:
```
data(r,c,nr,nc) = table.data;
```
That is just a call to `updBlock()` but is easier to read IMO. I wouldn't think the `updAsMatrix()` call is required.